### PR TITLE
Reduce regex use in CAST

### PR DESCRIPTION
**Pull request by andygrove**
This replaces four uses of regex in CAST with checks for expected length instead.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 1]
```diff
@@ -770,7 +769,7 @@ case class GpuCast(
     }
 
     /** This method does not close the `input` ColumnVector. */
-    def convertTimestampFullOr(
+    def convertFullTimestampOr(
         input: ColumnVector,
         orElse: ColumnVector): ColumnVector = {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:774]: This has the same issue of failing to close `orElse`if some earlier processing throws.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 2]
```diff
@@ -728,28 +733,22 @@ case class GpuCast(
...
         }
       }
     }
 
     /** This method does not close the `input` ColumnVector. */
-    def convertTimestampOr(
+    def convertVarLenTimestampOr(
         input: ColumnVector,
         regex: String,
         cudfFormat: String,
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:754]: Same `orElse` leaks on early throw here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 3]
```diff
@@ -783,9 +782,15 @@ case class GpuCast(
           isTimestamp1.or(isTimestamp2)
         }
       }
+
       val isValidTimestamp = withResource(isCudfMatch) { isCudfMatch =>
-        withResource(input.matchesRe(TIMESTAMP_REGEX_FULL)) { isRegexMatch =>
-          isCudfMatch.and(isRegexMatch)
+        val isValidLength = withResource(Scalar.fromInt(27)) { requiredLen =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:787]: It's not clear where `27` is coming from.  I suggest naming this constant as a `val` and placing it next to the format string(s) that it applies to.

**Review by jlowe — CHANGES_REQUESTED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Pull request merged**